lib/model/message_list.dart

```diff
@@ -493,11 +494,45 @@ class MessageListView with ChangeNotifier, _MessageSequence {
     notifyListeners();
   }
 
+  /// Update [narrow] for the result of a "with" narrow (topic permalink) fetch.
+  ///
+  /// To avoid an extra round trip, the server handles [ApiNarrowWith]
+  /// by returning results from the indicated message's current stream/topic
+  /// (if the user has access),
+  /// even if that differs from the narrow's stream/topic filters
+  /// because the message was moved.
+  ///
+  /// If such a "redirect" happened, this helper updates the stream and topic
+  /// in [narrow] to match the message's current conversation.
+  /// It also removes the "with" component from [narrow]
+  /// whether or not a redirect happened.
+  ///
+  /// See API doc:
+  ///   https://zulip.com/api/construct-narrow#message-ids
+  void _adjustTopicPermalinkNarrow(Narrow narrow_, Message? someFetchedMessageOrNull) {
```

Comment:
Looks like the caller passes the same `narrow` as the `narrow` that lives on the message list. Should we remove `narrow_` and just use `narrow` instead?

Reply:
I agree it's odd for this function to have a `narrow_` param; let's remove that.

It's convenient in the function's body to not keep using `narrow` each time we need it, because then we'd have to repeat things like `narrow as TopicNarrow` even in places where it's clear to us that the value can't be anything other than a `TopicNarrow`. I think what's happening there is: when we read `narrow`, the analyzer sees that as calling a getter, not accessing a variable, and it doesn't assume the getter will return the same value / `Narrow` subtype from one call to the next. And while in theory the analyzer could conclude that this class's `narrow` getter doesn't have that surprising behavior—it's just a field with the getter defined implicitly—it doesn't want to go and check for _subclasses_ that might override the getter with that surprising behavior.

How about a `narrow_` variable at the top of the function body, with a comment.

Reply:
Yeah, that explanation makes sense. I remember making the setter of `narrow` private might fix the issue, but I'm not sure. The proposed change sounds good to me too.

Reply:
Hmm nevermind, I think the local variable would be a better solution here.

See https://github.com/zulip/zulip-flutter/pull/808#discussion_r1677245004